etna/datasets/hierarchical_structure.py

```diff
@@ -0,0 +1,179 @@
+
+        self._find_graph_structure(level_structure)
+        hierarchy_levels = self._find_hierarchy_levels(level_structure)
+        tree_depth = len(hierarchy_levels)
+
+        if level_names is None:
+            level_names = [f"level_{i}" for i in range(tree_depth)]
+
+        if len(level_names) != tree_depth:
+            raise ValueError("Length of `level_names` must be equal to hierarchy tree depth!")
+
+        self.level_names = level_names
+        self._level_series: Dict[str, List[str]] = {level_names[i]: hierarchy_levels[i] for i in range(tree_depth)}
+        self._level_to_index: Dict[str, int] = {level_names[i]: i for i in range(tree_depth)}
+
+        self._segment_num_reachable_leafs: Dict[str, int] = self._get_num_reachable_leafs(hierarchy_levels)
+
+        self._segment_to_level: Dict[str, str] = {
+            segment: level for level in self._level_series for segment in self._level_series[level]
+        }
+
+    def _find_graph_structure(self, adj_list: Dict[str, List[str]]):
+        """Find hierarchy top level (root of tree)."""
+        children = set(chain(*adj_list.values()))
+        parents = set(adj_list.keys())
+
+        tree_roots = parents.difference(children)
+        if len(tree_roots) != 1:
+            raise ValueError("Invalid tree definition: unable to find root!")
+
+        hierarchy_interm_nodes = parents & children
+        hierarchy_leaves = children.difference(parents)
+
+        tree_root = tree_roots.pop()
+        self._hierarchy_root = tree_root
+
+        self._num_nodes = len(hierarchy_interm_nodes) + len(hierarchy_leaves) + 1
+
+    def _find_hierarchy_levels(self, hierarchy_structure: Dict[str, List[str]]) -> DefaultDict[int, List[str]]:
+        """Traverse hierarchy tree to group segments into levels."""
+        num_edges = sum(map(len, hierarchy_structure.values()))
+
+        if num_edges != self._num_nodes - 1:
+            raise ValueError("Invalid tree definition: invalid number of nodes and edges!")
+
+        leaves_levels = set()
+        levels = defaultdict(list)
+        seen_nodes = {self._hierarchy_root}
+        queue: Queue = Queue()
+        queue.put((self._hierarchy_root, 0))
+        while not queue.empty():
+            node, level = queue.get()
+            levels[level].append(node)
+            child_nodes = hierarchy_structure.get(node, [])
+
+            if len(child_nodes) == 0:
+                leaves_levels.add(level)
+
+            for adj_node in child_nodes:
+                queue.put((adj_node, level + 1))
+                seen_nodes.add(adj_node)
+
+        if len(seen_nodes) != self._num_nodes:
+            raise ValueError("Invalid tree definition: disconnected graph!")
+
+        if len(leaves_levels) != 1:
+            raise ValueError("All hierarchy tree leaves must be on the same level!")
+
+        return levels
+
+    def _get_num_reachable_leafs(self, hierarchy_levels: Dict[int, List[str]]) -> Dict[str, int]:
```

Comment:
We can remove argument `hierarchy_levels` as it is stored in `self`

Reply:
`hierarchy_levels` is not stored in `self`, because it is temporary structure.